app/api/Consignment.scala

```diff
@@ -0,0 +1,326 @@
+    (__ \ "address" \ "numberAndStreet").readNullable[String] and
+      (__ \ "address" \ "postalCode").readNullable[String] and
+      (__ \ "address" \ "city").readNullable[String]
+  ).tupled.map {
+    case (Some(streetAndNumber), postcode, Some(city)) =>
+      Some(AddressType01(streetAndNumber, postcode, city))
+    case _ => None
+  }
+
+}
+
+object postcodeAddressType02 {
+
+  implicit val reads: Reads[Option[PostcodeAddressType02]] = (
+    (__ \ "postalCode" \ "streetNumber").readNullable[String] and
+      (__ \ "postalCode" \ "postalCode").readNullable[String] and
+      (__ \ "postalCode" \ "country" \ "code").readNullable[String]
+  ).tupled.map {
+    case (streetAndNumber, Some(postcode), Some(country)) =>
+      Some(PostcodeAddressType02(streetAndNumber, postcode, country))
+    case _ => None
+  }
+
+}
+
+object contactPerson {
+
+  val name: String            = "name"
+  val telephoneNumber: String = "telephoneNumber"
+
+  def reads[T](apply: (String, String, Option[String]) => T): Reads[T] = (
+    (__ \ name).read[String] and
+      (__ \ telephoneNumber).read[String] and
+      None
+  )(apply)
+
+}
+
+object contactPersonType06 {
+
+  implicit val reads: Reads[ContactPersonType06] =
+    contactPerson.reads(ContactPersonType06)
+
+}
+
+object incidentType01 {
+
+  private lazy val convertIncidentCode: String => String = {
+    case "deviatedFromItinerary"         => "1"
+    case "sealsBrokenOrTampered"         => "2"
+    case "transferredToAnotherTransport" => "3"
+    case "partiallyOrFullyUnloaded"      => "4"
+    case "carrierUnableToComply"         => "5"
+    case "unexpectedlyChanged"           => "6"
+    case _                               => throw new Exception("Invalid incident code")
+  }
+
+  def reads(index: Int): Reads[IncidentType01] = (
+    (index.toString: Reads[String]) and
+      (__ \ "incidentCode").read[String].map(convertIncidentCode) and
+      (__ \ "incidentText").read[String] and
+      (__ \ "endorsement").readNullable[EndorsementType01](endorsementType01.reads) and
+      __.read[LocationType01](locationType01.reads) and
+      (__ \ "equipments").readArray[TransportEquipmentType01](transportEquipmentType01.reads) and
+      __.read[Option[TranshipmentType01]](transhipmentType01.reads)
+  ).apply {
+    (a, b, c, d, e, f, g) =>
+      IncidentType01(
+        sequenceNumber = a,
+        code = b,
+        text = c,
+        Endorsement = d,
+        Location = e,
+        TransportEquipment = f,
+        Transhipment = g
+      )
+  }
+
+}
+
+object endorsementType01 {
+
+  def reads: Reads[EndorsementType01] = (
+    (__ \ "date").read[String] and
+      (__ \ "authority").read[String] and
+      (__ \ "location").read[String] and
+      (__ \ "country" \ "code").read[String]
+  ).apply {
+    (a, b, c, d) =>
+      EndorsementType01(
+        date = a,
+        authority = b,
+        place = c,
+        country = d
+      )
+  }
```

Comment:
Could change this to `)(EndorsementType01.apply _)`

Reply:
Works if you provide the conversion

```
  def reads: Reads[EndorsementType01] = (
    (__ \ "date").read[String].map(stringToXMLGregorianCalendar) and
      (__ \ "authority").read[String] and
      (__ \ "location").read[String] and
      (__ \ "country" \ "code").read[String]
  )(EndorsementType01.apply _)
```